### PR TITLE
enh: print ArrayConstant in NumPy style

### LINKED ISSUE #3885: enh: add support to print ConstantArrays
https://github.com/lfortran/lfortran/pull/3859#discussion_r1567519497

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #3885.

### CODE REVIEW

**Comment by Pranavchiku** [comment]
Do we need to restrict printing upto two decimals or something like this? If yes, I'll make changes accordingly.

**Comment by Pranavchiku** [comment]
Here is failure:

I am unable to reproduce on my device, will try to reproduce on linux

```console
compiler_tester.tester.RunException: Testing with reference output failed.
../integration_tests/modules_39.f90 * asr
The JSON metadata differs against reference results
Reference JSON: tests/reference/asr-modules_39-89c7d0d.json
Output JSON:    tests/output/asr-modules_39-89c7d0d.json
Omitting 10 identical items
Differing items:
{'stdout_hash': 'da141e60d1412ea23fba37cf81c86274b67c12261d054a9b26077fa1'} != {'stdout_hash': 'a559bafc9605fdaef7458a0d5ab5e2d4b4a8fc7f13b8be77437741c0'}
Diff against: tests/reference/asr-modules_39-89c7d0d.stdout
113c113
<                                                 ["*"]
---
>                                                 ["*CiV"]

xonsh: For full traceback set: $XONSH_SHOW_TRACEBACK = True
Traceback (most recent call last):
  File "ci/test.xsh", line 44, in <module>
    ./run_tests.py
```

File: tests/reference/asr-array_02_pack-233bae0.stdout [hunk 1]
```diff
@@ -56,6 +56,7 @@
                         (Var 2 m)
                         (ArrayConstant
                             16
+                            [1, 0, 02]
```
Comment (gxyd) [inline tests/reference/asr-array_02_pack-233bae0.stdout:59]: missing a comma here?

File: tests/reference/asr-array_02_pack-233bae0.stdout [hunk 2]
```diff
@@ -151,6 +153,7 @@
                                     NotEq
                                     (ArrayConstant
                                         16
+                                        [1, 2, 34]
```
Comment (gxyd) [inline tests/reference/asr-array_02_pack-233bae0.stdout:156]: here as well I guess.

**Comment by gxyd** [comment]
On my machine (Macbook Air M1) on running integration_tests:
```bash
99% tests passed, 1 tests failed out of 1052

Label Time Summary:
c                 = 185.03 sec*proc (71 tests)
c_nopragma        =   8.17 sec*proc (3 tests)
cpp               = 113.82 sec*proc (45 tests)
fortran           = 267.85 sec*proc (103 tests)
gfortran          = 2793.38 sec*proc (1052 tests)
llvm              = 2340.27 sec*proc (882 tests)
llvm2             =  23.03 sec*proc (9 tests)
llvmImplicit      =  99.32 sec*proc (38 tests)
llvmStackArray    = 278.82 sec*proc (105 tests)
llvm_nopragma     =   8.17 sec*proc (3 tests)
llvm_rtlib        =   2.52 sec*proc (1 test)
llvm_wasm         = 2267.97 sec*proc (855 tests)
wasm              = 495.49 sec*proc (188 tests)
x86               =  45.30 sec*proc (19 tests)

Total Test time (real) = 350.45 sec

The following tests FAILED:
	475 - intrinsics_211 (Failed)
Errors while running CTest
Command failed.
```

**Comment by gxyd** [comment]
I haven't given any proper review, just a few things I noticed on a quick glance.

File: tests/reference/asr-array_02_pack-233bae0.stdout (hunk 1 above)
Comment (Pranavchiku) [reply tests/reference/asr-array_02_pack-233bae0.stdout:59]: Ahh thanks for pointing, I see why this happens.

**Comment by Pranavchiku** [comment]
That happens in main as well, opened https://github.com/lfortran/lfortran/issues/3893 for it.

**Review by certik — APPROVED** [review]
Beautiful, thank you. This is exactly what I wanted. I think this provides a very solid foundation in ASR to represent compile time arrays and to easily see what's in it.

**Comment by Thirumalai-Shaktivel** [comment]
Were you able to reproduce the Ubuntu issue?
I think the issue is with the `inline std::string fetch_ArrayConstant_value(void *data, ASR::ttype_t* type, int i) {`: `case ASR::ttypeType::Character`, see what's happening there.

**Comment by Pranavchiku** [comment]
Somehow in `set_ArrayConstant_data`, things get wrong, I'll debug it.

**Comment by Pranavchiku** [comment]
To fix this I think we'll have to null terminate the string

**Comment by Pranavchiku** [comment]
I think latest commit shall do the work.

**Pull request merged**